Tail.scala

```diff
@@ -0,0 +1,226 @@
+                                                        // sinusoid frequency should be a multiple of n = 8, samples should cover at least 2 periods
+    io.signalIn.ready := DeserializeModule.in.ready
+
+    // Connect Deserialize to FFT
+    for (j <- 0 until params.lanes) FFTModule.in.bits(j) := DeserializeModule.out.bits(j)
+    FFTModule.in.valid := DeserializeModule.out.valid
+    FFTModule.in.sync := DeserializeModule.out.sync // todo ask yue: what does sync do? -- not used but leave it in assigned to false in Deserialize
+
+    // Connect FFT to Unscramble
+    /* Animesh: FFT outputs values with bits reversed (Ex: input of 001 becomes output of 100)
+       Unscramble fixes the bit order
+    */
+    for (j <- 0 until params.lanes) UnscrambleModule.in.bits(j) := FFTModule.out.bits(j)
+    UnscrambleModule.in.valid := FFTModule.out.valid
+    UnscrambleModule.in.sync := FFTModule.out.sync
+
+    for (j <- 0 until params.lanes) io.signalOut.bits(j) := UnscrambleModule.out.bits(j)
+    io.signalOut.valid := UnscrambleModule.out.valid
+    UnscrambleModule.out.ready := io.signalOut.ready
+}
+
+class LazyTail(val config: FixedTailParams)(implicit p: Parameters) extends LazyModule {
+  val device = new SimpleDevice("fft-generator", Seq("cpu")) // add an entry to the DeviceTree in the BootROM so that it can be read by a Linux driver (9.2 chipyard docs)
+  val node = TLRegisterNode(
+    address = Seq(AddressSet(p(FFTBaseAddrKey), 0xff)), // (base address + size) of regmap
+    device = device,
+    beatBytes = 8, // specifies interface width in bytes -- since we're connected to a 64bit bus, want an 8byte width (default is 4)
+    concurrency = 1 // size of the internal queue for TileLink requests, must be >0 for decoupled requests and responses (9.4 chipyard docs)
+  )
+
+   lazy val module = new LazyModuleImp(this) {
+    val tail = Module(new Tail(config))
+
+    val inputWire = Wire(Decoupled(UInt((config.IOWidth * 2).W))) // protoIn defines the input points, are 16 bits
+    inputWire.ready := tail.io.signalIn.ready // WriteQueueBlock.scala 67
+    tail.io.signalIn.bits := inputWire.bits.asTypeOf(config.protoIn) // WriteQueueBlock.scala 66
+    tail.io.signalIn.valid := inputWire.valid
+
+    var outputRegs = new ListBuffer[chisel3.UInt]() // todo ask abe: Is this the right type annotation?? Is it UInt or Regsmth
+    for (i <- 0 until config.n) { // todo parameterize
+      outputRegs += RegEnable(tail.io.signalOut.bits(i).asUInt(), 0.U, tail.io.signalOut.valid)
+    }
+
+    var regMap = new ListBuffer[(Int, Seq[freechips.rocketchip.regmapper.RegField])]()
+    regMap += (0x00 -> Seq(RegField.w(config.IOWidth * 2, inputWire)))
+
+    // var regmapSeq: Seq[(Int, Seq[RegField])] = Seq( (0x00, Seq(RegField.w(16, inputWire))) )
+    for (i <- 0 until config.n) { // todo parameterize
+      regMap += (0x00 + (i+1) * 8 -> Seq(RegField.r(config.IOWidth * 2, outputRegs(i))))
+    }
+
```

Comment:
I feel like having a `for` loop going through this is unnecessary, can you see if there is a way to make this less verbose?